include/grpcpp/impl/codegen/client_callback.h

```diff
@@ -90,6 +90,149 @@ class CallbackUnaryCallImpl {
 };
 
 }  // namespace internal
+
+/// TODO(vjpai): Move these contents out of experimental
+/// TODO(vjpai): Merge with contents of sync_stream.h or async_stream.h if
+/// possible, including getting their interfaces into the interface classes
+namespace experimental {
+
+// Forward declaration
+template <class W, class R>
+class ClientCallbackReaderWriter;
+
+/// TODO(vjpai): Put this factory into ::grpc::internal when possible
+template <class W, class R>
+class ClientCallbackReaderWriterFactory {
+ public:
+  static ::grpc::experimental::ClientCallbackReaderWriter<W, R>* Create(
+      ::grpc::ChannelInterface* channel,
+      const ::grpc::internal::RpcMethod& method, ClientContext* context) {
+    ::grpc::internal::Call call =
+        channel->CreateCall(method, context, channel->CallbackCQ());
+    return new (g_core_codegen_interface->grpc_call_arena_alloc(
+        call.call(),
+        sizeof(::grpc::experimental::ClientCallbackReaderWriter<W, R>)))::grpc::
+        experimental::ClientCallbackReaderWriter<W, R>(call, context);
+  }
+};
+
+/// Callback-based client-side API for bi-directional streaming RPCs,
+/// where the outgoing message stream coming from the client has messages of
+/// type \a W, and the incoming messages stream coming from the server has
+/// messages of type \a R.
+/// TODO(vjpai): make this a derived class of an interface class
+template <class W, class R>
+class ClientCallbackReaderWriter final {
+ public:
+  // always allocated against a call arena, no memory free required
+  static void operator delete(void* ptr, std::size_t size) {
+    assert(size == sizeof(ClientCallbackReaderWriter));
+  }
+
+  // This operator should never be called as the memory should be freed as part
+  // of the arena destruction. It only exists to provide a matching operator
+  // delete to the operator new so that some compilers will not complain (see
+  // https://github.com/grpc/grpc/issues/11301) Note at the time of adding this
+  // there are no tests catching the compiler warning.
+  static void operator delete(void*, void*) { assert(0); }
+
+  void StartCall(std::function<void(Status)> on_completion) {
+    StartCall(std::move(on_completion), nullptr);
+  }
+
+  void StartCall(std::function<void(Status)> on_completion,
+                 std::function<void(bool)> on_metadata_available) {
+    // This call may initiate two batches
+    // 1. Send initial metadata/recv initial metadata, optional callback
+    // 2. Recv trailing metadata, on_completion callback
+
```

Comment:
Why do we have a set of Start ops and Finish ops. What would happen if we batch the Send Initial Metadata, Recv Initial Metadata and Receive Status on client all in one op?

Reply:
I think you might be right in at least some cases. If the user wants the recv initial metadata callback, then they shouldn't be lumped together (since that would delay the callback until the entire batch is done). If the user doesn't want a separate callback, though, it should be fine to go ahead and lump them all together since the important concern is to make sure that we're at least initiating the recv initial metadata before the recv status or any recv messages.

Reply:
I have a dumber follow up question. Why do we need a recv status when we already have a recv initial metadata? Wouldn't receiving the intial metadata already signify that the server is alive.

Reply:
Following our conversation: different purposes, status comes at the end and signifies the successful/unsuccessful _completion_ of the RPC, not just the condition of the server.